File: 990.等式方程的可满足性.py
#
# @lc app=leetcode.cn id=990 lang=python3
#
# [990] 等式方程的可满足性
#

# @lc code=start
class Solution:
    def equationsPossible(self, equations: List[str]) -> bool:
        class UF:
            def __init__(self, n) -> None:
                self.parent = list(range(n)) # 指向自己
                self.size = [1]*n # 每个树都仅有自己这个节点，为了平衡引入该变量
            
            def find(self, x):
                # 在find里做了修改操作，其实不该出现在这
                if self.parent[x] != x:
                    self.parent[x] = self.find(self.parent[x]) # 将其父节点设为其祖父节点
                return self.parent[x]
            
            def union(self, x, y): # 这里可以计算连通分量
                root_x, root_y = self.find(x), self.find(y)
                if root_x == root_y: return
                if self.size[root_x] < self.size[root_y]:
                    root_x, root_y = root_y, root_x
                self.parent[root_y] = root_x
                self.size[root_x] += self.size[root_y]
            
            def connected(self, x, y):
                return self.find(x) == self.find(y)
        uf = UF(26)
        for eq in equations:
            if eq[1] == '=':
                x, y = eq[0], eq[3]
                uf.union(ord(x)-ord('a'), ord(y)-ord('a'))
        for eq in equations:
            if eq[1] == '!':
                x, y = eq[0], eq[3]
                if uf.connected(ord(x)-ord('a'),ord(y)-ord('a')):
                    return False
        return True

# @lc code=end

File: 990.等式方程的可满足性.java
/*
 * @lc app=leetcode.cn id=990 lang=java
 *
 * [990] 等式方程的可满足性
 */

// @lc code=start
class Solution {
    public boolean equationsPossible(String[] equations) {
        UF uf = new UF(26);
        //先把等于的聚类
        for(String eq:equations){
            if(eq.charAt(1)=='='){
                int a = (int)eq.charAt(0) - (int)'a';
                int b = (int)eq.charAt(3) - (int)'a';
                uf.union(a, b);
            }
        }
        for(String eq:equations){
            if(eq.charAt(1)=='!'){
                int a = (int)eq.charAt(0) - (int)'a';
                int b = (int)eq.charAt(3) - (int)'a';
                if(uf.connected(a, b))return false;
            }
        }
        return true;
    }
}
class UF{
    private int[] parent;
    private int[] size;

    public UF(int n){
        parent = new int[n];
        size = new int[n];
        for(int i=0;i<n;i++){
            parent[i] = i;
        }
        for(int i=0;i<n;i++){
            size[i] = 1;
        }
    }

    public int find(int x){
        if(parent[x]!=x)parent[x] = find(parent[x]);
        return parent[x];
    }

    public void union(int x,int y){
        int root_x = find(x);
        int root_y = find(y);
        if(root_x==root_y)return;
        if(size[root_x]<size[root_y]){
            int tmp = root_x;
            root_x = root_y;
            root_y = tmp;
        }
        parent[root_y] = root_x;
        size[root_x] += size[root_y];
    }

    public boolean connected(int x, int y){
        return find(x) == find(y);
    }

}
// @lc code=end

